### PR TITLE
Make all Nodes immutable but still easy to "modify"

### PR DESCRIPTION
**Pull request by andimarek**
The goal of this PR to make all Nodes immutable, but "modifying" the Ast should be still easy to.

This means that you can easily get a new tree when you modify a Node, even if the Node is deep inside the tree.

contains contribution from @felipe-gdr :-)

### CODE REVIEW

**Comment by andimarek** [comment]
There still might be some improvements, but I expect that all major parts for that change are in place:

Every Node is now immutable. 

It introduces the concept of a Zipper: basically a pointer to a specific Node in a immutable Tree which also keeps track of where where the current Node is in respect to the parent. (see [here](https://github.com/graphql-java/graphql-java/blob/5f11a736b4aa2abbe99826ff6de23eceeeeed03a/src/main/java/graphql/language/AstZipper.java) )

Some background details: the problem with an immutable Ast is, that it is very easy to change a Node and all the children, but is very hard then also to change the parents referencing the node you just changed, because a Node just knows about its children, nothing about it parent. Zipper is a general concept which helps with that. I can recommend http://learnyouahaskell.com/zippers as a more detailed explanation. I also took inspiration in implementing it from this book. (Warning: it actually talks about Haskell and not Java)

Additionally there is the class [AstMultiZipper](https://github.com/graphql-java/graphql-java/blob/5f11a736b4aa2abbe99826ff6de23eceeeeed03a/src/main/java/graphql/language/AstMultiZipper.java) which is a list of Zippers and allows keeping track of multiple changes in a Tree at the same time.

`AstTransformer` now allows for modifying the Ast (`Document`) and getting a new Tree back in a very convenient way by leveraging `AstMultiZipper`. See [here](https://github.com/graphql-java/graphql-java/blob/5f11a736b4aa2abbe99826ff6de23eceeeeed03a/src/test/groovy/graphql/language/AstTransformerTest.groovy) for a working test case how it looks in detail.
 
Feedback welcome.

Also: Now that every Node is immutable you can't change a query anymore at any point. If you have some use cases which requires changing the query after it was create, please comment here. We wanna offer specific hooks where a query can be modified.

**Comment by andimarek** [comment]
Ping @tinnou for feedback. thanks

File: src/main/java/graphql/execution2/result/ExecutionResultMultiZipper.java [hunk 1]
```diff
@@ -85,6 +87,7 @@ private ExecutionResultZipper moveUp(ExecutionResultNode parent, List<ExecutionR
     }
```
Comment (bbakerman) [inline src/main/java/graphql/execution2/result/ExecutionResultMultiZipper.java]: TODO ^

File: src/main/java/graphql/execution2/result/ExecutionResultMultiZipper.java [hunk 2]
```diff
@@ -85,6 +87,7 @@ private ExecutionResultZipper moveUp(ExecutionResultNode parent, List<ExecutionR
...
 
     private Map<ExecutionResultNode, List<ExecutionResultZipper>> zipperWithSameParent(List<ExecutionResultZipper> zippers) {
+        //todo: change to LinkedHashMap
         return zippers.stream().collect(Collectors.groupingBy(ExecutionResultZipper::getParent));
     }
```
Comment (bbakerman) [inline src/main/java/graphql/execution2/result/ExecutionResultMultiZipper.java]: TODO ^

I think you put in a supplier of the Map

File: src/main/java/graphql/language/AstMultiZipper.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+            List<AstZipper> newZippers = new ArrayList<>();
+            for (Map.Entry<Node, List<AstZipper>> entry : sameParent.entrySet()) {
+                AstZipper newZipper = moveUp(entry.getKey(), entry.getValue());
+                newZippers.add(newZipper);
+            }
+            curZippers.removeAll(deepestZippers);
+            curZippers.addAll(newZippers);
+        }
+        assertTrue(curZippers.size() == 1, "illegal state");
```
Comment (bbakerman) [inline src/main/java/graphql/language/AstMultiZipper.java]: better assertion message - eg state the precondition better or have no message

File: src/main/java/graphql/language/Directive.java [hunk 4]
```diff
@@ -18,6 +18,8 @@
     private final String name;
     private final List<Argument> arguments = new ArrayList<>();
 
+    private static final String CHILD_ARGUMENTS = "arguments";
```
Comment (bbakerman) [inline src/main/java/graphql/language/Directive.java]: not sure on the CHILD_ smurfing

The thing that represents "arguments" could be called arguments

File: src/main/java/graphql/language/FloatValue.java [hunk 5]
```diff
@@ -40,6 +40,20 @@ public BigDecimal getValue() {
...
+        return NodeChildrenContainer.newNodeChildrenContainer().build();
+    }
+
+    @Override
+    public FloatValue withNewChildren(NodeChildrenContainer newChildren) {
+        if (!newChildren.isEmpty()) {
+            throw new IllegalArgumentException("Cannot pass non-empty newChildren to Node that doesn't hold children");
+        }
+
+        return this;
```
Comment (bbakerman) [inline src/main/java/graphql/language/FloatValue.java]: This one does not seem to be finished.

Is this a BUG?

File: src/main/java/graphql/language/NullValue.java [hunk 6]
```diff
@@ -25,6 +25,20 @@ protected NullValue(SourceLocation sourceLocation, List<Comment> comments, Ignor
...
+    }
+
+    @Override
+    public NullValue withNewChildren(NodeChildrenContainer newChildren) {
+        if (!newChildren.isEmpty()) {
+            throw new IllegalArgumentException("Cannot pass non-empty newChildren to Node that doesn't hold children");
+        }
+
+        return this;
+    }
```
Comment (bbakerman) [inline src/main/java/graphql/language/NullValue.java]: Yeah - defintely common method if this is needed

    @Override
    public NullValue withNewChildren(NodeChildrenContainer newChildren) {
           return NodeChildrenContainer.assertNotEmpty(newChildren);
    }

File: src/main/java/graphql/util/DefaultTraverserContext.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+    @Override
+    public Object getSharedContextData() {
+        return sharedContextData;
+    }
+
+    /**
+     * PRIVATE: Used by {@link Traverser}
+     */
+    void setCurAccValue(Object curAccValue) {
```
Comment (bbakerman) [inline src/main/java/graphql/util/DefaultTraverserContext.java]: bad Javadoc - its going to lead to warnings and set of my OCD on having a clean command line compile / build

File: src/main/java/graphql/util/NodePosition.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        NodePosition that = (NodePosition) o;
+        return index == that.index &&
+                Objects.equals(name, that.name);
```
Comment (bbakerman) [inline src/main/java/graphql/util/NodePosition.java]: Objects.equals() ??

File: src/test/groovy/graphql/language/AstMultiZipperTest.groovy [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+    def "toRootNode with one zipper"() {
+        def rootNode = TestUtil.parseQuery("{root{midA{leafA leafB} midB{leafC leafD}}}")
+        def operationDefinition = rootNode.children[0] as OperationDefinition
+
+        def rootField = operationDefinition.selectionSet.children[0] as Field
+        def midA = rootField.selectionSet.children[0] as Field
+        def leafA = midA.selectionSet.children[0] as Field
+        def leafB = midA.selectionSet.children[0] as Field
+
```
Comment (bbakerman) [inline src/test/groovy/graphql/language/AstMultiZipperTest.groovy]: I think the indices here are wrong.

     def leafB = midA.selectionSet.children[1] as Field

here and below

File: src/test/groovy/graphql/language/AstMultiZipperTest.groovy [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        def midB = rootField.selectionSet.children[1] as Field
+        def leafC = midB.selectionSet.children[0] as Field
+        def leafD = midB.selectionSet.children[0] as Field
```
Comment (bbakerman) [inline src/test/groovy/graphql/language/AstMultiZipperTest.groovy]: indices are wrong I think

File: src/main/java/graphql/execution2/result/ExecutionResultMultiZipper.java (hunk 1 above)
Comment (andimarek) [reply src/main/java/graphql/execution2/result/ExecutionResultMultiZipper.java]: done

File: src/main/java/graphql/execution2/result/ExecutionResultMultiZipper.java (hunk 2 above)
Comment (andimarek) [reply src/main/java/graphql/execution2/result/ExecutionResultMultiZipper.java]: done

File: src/main/java/graphql/language/FloatValue.java (hunk 5 above)
Comment (andimarek) [reply src/main/java/graphql/language/FloatValue.java]: well: you can't add children to a FloatValue, so we only allow empty children here

File: src/main/java/graphql/util/DefaultTraverserContext.java (hunk 7 above)
Comment (andimarek) [reply src/main/java/graphql/util/DefaultTraverserContext.java]: changed it no java doc

File: src/test/groovy/graphql/language/AstMultiZipperTest.groovy (hunk 9 above)
Comment (andimarek) [reply src/test/groovy/graphql/language/AstMultiZipperTest.groovy]: fixed

File: src/test/groovy/graphql/language/AstMultiZipperTest.groovy (hunk 10 above)
Comment (andimarek) [reply src/test/groovy/graphql/language/AstMultiZipperTest.groovy]: fixed

File: src/main/java/graphql/language/NullValue.java (hunk 6 above)
Comment (andimarek) [reply src/main/java/graphql/language/NullValue.java]: refactored it

File: src/main/java/graphql/execution2/result/ExecutionResultMultiZipper.java [hunk 11]
```diff
@@ -68,7 +73,7 @@ public ExecutionResultMultiZipper withReplacedZipper(ExecutionResultZipper oldZi
 
     private List<ExecutionResultZipper> getDeepestZippers(List<ExecutionResultZipper> zippers) {
         Map<Integer, List<ExecutionResultZipper>> grouped =
-                zippers.stream().collect(Collectors.groupingBy(executionResultZipper -> executionResultZipper.getBreadcrumbList().size()));
+                zippers.stream().collect(Collectors.groupingBy(executionResultZipper -> executionResultZipper.getBreadcrumbList().size(), LinkedHashMap::new, mapping(Function.identity(), toList())));
         Integer maxLevel = Collections.max(grouped.keySet());
```
Comment (bbakerman) [inline src/main/java/graphql/execution2/result/ExecutionResultMultiZipper.java]: Almost deserves it is on `Collector` based puirely on how much boilerplate.  Perhaps its own method

`zippers.stream().collect(Collectors.groupingBy(thisGrouping(executionResultZipper -> executionResultZipper.getBreadcrumbList().size()));`

File: src/main/java/graphql/execution2/result/ExecutionResultMultiZipper.java (hunk 2 above)
Comment (bbakerman) [reply src/main/java/graphql/execution2/result/ExecutionResultMultiZipper.java]: Yeah I reckon create its own helper method

File: src/main/java/graphql/language/AstZipper.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
+package graphql.language;
```
Comment (bbakerman) [inline src/main/java/graphql/language/AstZipper.java]: Should all this transformation code be in `graphql.language` or one down in say `graphql.language.transformation` ?

I think the latter

File: src/main/java/graphql/language/AstZipper.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        return new AstZipper(curNode, newBreadcrumbs);
+    }
+
+    public AstZipper changeLocation(NodeLocation newLocationInParent, List<AstBreadcrumb> newBreadcrumbs) {
+        // validate position
+        return new AstZipper(curNode, newBreadcrumbs);
+    }
+
+    public AstZipper moveUp() {
+        return null;
```
Comment (bbakerman) [inline src/main/java/graphql/language/AstZipper.java]: is this not done???

Comment (bbakerman) [reply src/main/java/graphql/language/AstZipper.java]: and hence not tested??

File: src/main/java/graphql/language/FloatValue.java (hunk 5 above)
Comment (bbakerman) [reply src/main/java/graphql/language/FloatValue.java]: I would create a helper method for these checks

```
@Override
    public FloatValue withNewChildren(NodeChildrenContainer newChildren) {
        NodeChildrenContainer.assertEmpty(newChildren);
        return this;
    }
```

File: src/main/java/graphql/language/IntValue.java [hunk 14]
```diff
@@ -40,6 +42,17 @@ public BigInteger getValue() {
...
     }
 
+    @Override
+    public NodeChildrenContainer getNamedChildren() {
+        return NodeChildrenContainer.newNodeChildrenContainer().build();
+    }
+
+    @Override
+    public IntValue withNewChildren(NodeChildrenContainer newChildren) {
+        assertNewChildrenAreEmpty(newChildren);
```
Comment (bbakerman) [inline src/main/java/graphql/language/IntValue.java]: Are you have one.  Do it on FLoatValue

File: src/main/java/graphql/language/Node.java [hunk 15]
```diff
@@ -24,6 +24,10 @@
      */
     List<Node> getChildren();
 
+    NodeChildrenContainer getNamedChildren();
```
Comment (bbakerman) [inline src/main/java/graphql/language/Node.java]: JavaDoc

File: src/main/java/graphql/language/NodeChildrenContainer.java [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+import static graphql.Assert.assertNotNull;
+
+@PublicApi
+public class NodeChildrenContainer {
```
Comment (bbakerman) [inline src/main/java/graphql/language/NodeChildrenContainer.java]: JavaDoc?

Maybe a seperate PR for the JavaDoc say

File: src/test/groovy/graphql/util/TraverserTest.groovy [hunk 17]
```diff
@@ -327,10 +347,106 @@ class TraverserTest extends Specification {
...
+                    def curAcc = context.getCurrentAccumulate()
+                    if (context.getPosition() != null) {
+                        curAcc.add(context.getPosition())
+                    }
+                    TraversalControl.CONTINUE
+                },
+                leave: { TraverserContext context ->
+                    TraversalControl.CONTINUE
+                }
+        ] as TraverserVisitor
```
Comment (bbakerman) [inline src/test/groovy/graphql/util/TraverserTest.groovy]: what is this crazy syntax?? I didnt know you could do this! :)

Comment (andimarek) [reply src/test/groovy/graphql/util/TraverserTest.groovy]: a groovy map with closures/functions as value ... not so crazy, or?

File: src/main/java/graphql/language/AstBreadcrumb.java [hunk 18]
```diff
@@ -0,0 +1,53 @@
+package graphql.language;
+
+import graphql.PublicApi;
+
+import java.util.Objects;
+
+@PublicApi
+public class AstBreadcrumb {
```
[the remark this answers is not in the document]
Comment (andimarek) [reply src/main/java/graphql/language/AstBreadcrumb.java]: added

File: src/main/java/graphql/language/AstMultiZipper.java [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static graphql.Assert.assertTrue;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+import static java.util.stream.Collectors.toList;
+
+@PublicApi
+public class AstMultiZipper {
```
[the remark this answers is not in the document]
Comment (andimarek) [reply src/main/java/graphql/language/AstMultiZipper.java]: added

File: src/main/java/graphql/language/Node.java (hunk 15 above)
Comment (andimarek) [reply src/main/java/graphql/language/Node.java]: added

File: src/main/java/graphql/language/NodeChildrenContainer.java (hunk 16 above)
Comment (andimarek) [reply src/main/java/graphql/language/NodeChildrenContainer.java]: added

File: src/main/java/graphql/language/AstMultiZipper.java (hunk 3 above)
Comment (andimarek) [reply src/main/java/graphql/language/AstMultiZipper.java]: done

File: src/main/java/graphql/language/FloatValue.java (hunk 5 above)
Comment (andimarek) [reply src/main/java/graphql/language/FloatValue.java]: done via helper method

File: src/main/java/graphql/language/Argument.java [hunk 20]
```diff
@@ -42,21 +44,26 @@ public Value getValue() {
...
         result.add(value);
         return result;
     }
 
+    @Override
+    public NodeChildrenContainer getNamedChildren() {
+        return NodeChildrenContainer.newNodeChildrenContainer()
+                .child(CHILD_VALUE, value)
+                .build();
+    }
```
[the remark this answers is not in the document]
Comment (andimarek) [reply src/main/java/graphql/language/Argument.java]: done

**Comment by kaqqao** [comment]
> Now that every Node is immutable you can't change a query anymore at any point. If you have some use cases which requires changing the query after it was create, please comment here. 

Perhaps I'm too late, but I'm aware of users that allow queries for objects without a subselection, by modifying the query on the fly to include all the direct fields in such cases. This was done via an instrumentation.
Is this use case still possible?

**Comment by kaqqao** [comment]
Ah, just seen #1345 which caters exactly to this use-case :)

**Pull request merged**